### PR TITLE
Update granite.md

### PR DESCRIPTION
**Pull request by tanuj-rai**
# What does this PR do?

#36979,
 This PR updates the model-card for the Granite model as described in https://github.com/huggingface/transformers/issues/36979,

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@stevhliu

### CODE REVIEW

File: docs/source/en/model_doc/granite.md [hunk 1]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
 <div class="flex flex-wrap space-x-1">
 <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">
 <img alt="FlashAttention" src="https://img.shields.io/badge/%E2%9A%A1%EF%B8%8E%20FlashAttention-eae0c8?style=flat">
 <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
 </div>
 
-## Overview
+# Granite
+
+[Granite](https://huggingface.co/ibm-granite) model was proposed in [Power Scheduler: A Batch Size and Token Number Agnostic Learning Rate Scheduler](https://arxiv.org/abs/2408.13359) by Yikang Shen, Matthew Stallone, Mayank Mishra, Gaoyuan Zhang, Shawn Tan, Aditya Prasad, Adriana Meza Soria, David D. Cox, and Rameswar Panda.
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:26]:
```suggestion
[Granite](https://huggingface.co/papers/2408.13359) is a 3B parameter language model trained with the Power scheduler. Discovering a good learning rate for pretraining large language models is difficult because it depends on so many variables (batch size, number of training tokens, etc.) and it is expensive to perform a hyperparameter search. The Power scheduler is based on a power-law relationship between the variables and their transferability to larger models. Combining the Power scheduler with Maximum Update Parameterization (MUP) allows a model to be pretrained with one set of hyperparameters regardless of all the variables.
```

File: docs/source/en/model_doc/granite.md [hunk 2]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
+Granite model is a series of AI language models built for businesses, not just for general use. It is a super-focused tool for enterprise tasks like summarizing reports, answering questions, or generating code. It is trained on curated datasets (think finance, legal, code, and academic stuff) and filters out junk like duplicates or harmful content with their HAP detector. It comes in different sizes, from super lightweight (sub-billion parameters) up to 34 billion, so you can pick what fits your needs.
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:27]:
```suggestion
```

File: docs/source/en/model_doc/granite.md [hunk 3]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
+
+You can find all the original Granite checkpoints under the [Granite](https://huggingface.co/ibm-granite) collection.
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:29]:
```suggestion
You can find all the original Granite checkpoints under the [IBM-Granite](https://huggingface.co/ibm-granite) organization.
```

File: docs/source/en/model_doc/granite.md [hunk 4]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
+
+> [!TIP]
+> Click on the Granite models in the right sidebar for more examples of how to apply Granite to different language tasks.
 
-The Granite model was proposed in [Power Scheduler: A Batch Size and Token Number Agnostic Learning Rate Scheduler](https://arxiv.org/abs/2408.13359) by Yikang Shen, Matthew Stallone, Mayank Mishra, Gaoyuan Zhang, Shawn Tan, Aditya Prasad, Adriana Meza Soria, David D. Cox and Rameswar Panda.
+The example below demonstrates how to generate text based on an image with [`Pipeline`] or the [`AutoModel`] class.
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:34]:
```suggestion
The example below demonstrates how to generate text with [`Pipeline`], [`AutoModel`, and from the command line.
```

File: docs/source/en/model_doc/granite.md [hunk 5]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
 
-PowerLM-3B is a 3B state-of-the-art small language model trained with the Power learning rate scheduler. It is trained on a wide range of open-source and synthetic datasets with permissive licenses. PowerLM-3B has shown promising results compared to other models in the size categories across various benchmarks, including natural language multi-choices, code generation, and math reasoning.
+<hfoptions id="usage">
+<hfoption id="Pipeline">
 
-The abstract from the paper is the following:
+```py
+from transformers import pipeline
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:40]:
```suggestion
import torch
from transformers import pipeline
```

File: docs/source/en/model_doc/granite.md [hunk 6]
```diff
@@ -57,14 +76,42 @@ input_tokens = tokenizer(prompt, return_tensors="pt")
...
 # decode output tokens into text
 output = tokenizer.batch_decode(output)
-# loop over the batch to print, in this example the batch size is 1
+# loop over the batch to print, in this example, the batch size is 1
 for i in output:
     print(i)
 ```
-
 This model was contributed by [mayank-mishra](https://huggingface.co/mayank-mishra).
+</hfoption>
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:84]: Add another `<hfoption>` for the `transformers-cli` example:

```bash
echo -e "Explain quantum computing simply." | transformers-cli run --task text-generation --model ibm-granite/granite-3.3-8b-instruct --device 0
```

File: docs/source/en/model_doc/granite.md [hunk 7]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
-Tips:
+)
+
+prompt = "Explain quantum computing in simple terms:"
+output = pipe(prompt, max_new_tokens=200, do_sample=True, temperature=0.7)
+print(output[0]["generated_text"])
+```
+
+</hfoption>
+<hfoption id="AutoModel">
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:57]: Replace the current with this example:

```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("ibm-granite/granite-3.3-2b-base")
model = AutoModelForCausalLM.from_pretrained("ibm-granite/granite-3.3-2b-base", torch_dtype=torch.bfloat16, device_map="auto", attn_implementation="sdpa")

inputs = tokenizer("Explain quantum computing in simple terms", return_tensors="pt").to("cuda")
outputs = model.generate(**inputs, max_length=50, cache_implementation="static")
print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

File: docs/source/en/model_doc/granite.md [hunk 8]
```diff
@@ -9,33 +9,52 @@ Unless required by applicable law or agreed to in writing, software distributed
...
-In this paper, we study the correlation between optimal learning rate, batch size, and number of training tokens for the recently proposed WSD scheduler. After thousands of small experiments, we found a power-law relationship between variables and demonstrated its transferability across model sizes. Based on the observation, we propose a new learning rate scheduler, Power scheduler, that is agnostic about the number of training tokens and batch size. The experiment shows that combining the Power scheduler with Maximum Update Parameterization (\mup) can consistently achieve impressive performance with one set of hyperparameters regardless of the number of training tokens, batch size, model size, and even model architecture. Our 3B dense and MoE models trained with the Power scheduler achieve comparable performance as state-of-the-art small language models.
-We [open source](https://huggingface.co/collections/ibm/power-lm-66be64ae647ddf11b9808000) these pretrained models.*
+pipe = pipeline(
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:42]: Simplify this example to:

```py
import torch
from transformers import pipeline

pipeline = pipeline(
    task="text-generation",
    model="ibm-granite/granite-3.3-2b-base",
    torch_dtype=torch.bfloat16,
    device=0
)
pipeline("Explain quantum computing in simple terms ", max_new_tokens=50)
```

File: docs/source/en/model_doc/granite.md [hunk 9]
```diff
@@ -57,14 +76,42 @@ input_tokens = tokenizer(prompt, return_tensors="pt")
...
 This model was contributed by [mayank-mishra](https://huggingface.co/mayank-mishra).
+</hfoption>
+</hfoptions>
+
+Quantization reduces the memory burden of large models by representing the weights in a lower precision. Refer to the [Quantization](../quantization/overview) overview for more available quantization backends.
+
+The example below uses [bitsandbytes](../quantization/bitsandbytes) to only quantize the weights to int4.
+
+```python
+from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:92]: Wrong model checkpoint here ;)

```suggestion
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

quantization_config = BitsAndBytesConfig(load_in_4bit=True)
tokenizer = AutoTokenizer.from_pretrained("ibm-granite/granite-3.3-8b-base")
model = AutoModelForCausalLM.from_pretrained("ibm-granite/granite-3.3-8b-base", torch_dtype=torch.bfloat16, device_map="auto", attn_implementation="sdpa", quantization_config=quantization_config)

inputs = tokenizer("Explain quantum computing in simple terms", return_tensors="pt").to("cuda")
outputs = model.generate(**inputs, max_length=50, cache_implementation="static")
print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

File: docs/source/en/model_doc/granite.md [hunk 10]
```diff
@@ -57,14 +76,42 @@ input_tokens = tokenizer(prompt, return_tensors="pt")
...
+
+outputs = model.generate(**input_ids, max_new_tokens=32)
+print(tokenizer.decode(outputs[0]))
+```
 
+## Notes
 
+- Granite models support context lengths up to 128K tokens (e.g., in Granite-3.3), thanks to Rotary Position Embeddings     (RoPE). For long-context tasks like document analysis, ensure your input sequences are padded appropriately to avoid      attention issues.
+- Don’t use the torch_dtype parameter in from_pretrained() if you’re using FlashAttention-2, as it only supports fp16 or    bf16. Granite models are optimized for mixed precision, so use Automatic Mixed Precision; set fp16 or bf16 to True if 
+  using Trainer, or use torch.autocast.
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:113]: You can remove

**Review by stevhliu — COMMENTED** [review]
Thanks for your PR!

File: docs/source/en/model_doc/granite.md (hunk 7 above)
Comment (Tanuj-rai) [reply docs/source/en/model_doc/granite.md:58]: "Thank you @stevhliu for clarifying the AutoModel example. Earlier, I was unsure whether I should remove the example contributed by Mayank Mishra or leave it as is."

**Comment by Tanuj-rai** [comment]
Hey @stevhliu, I’ve made the changes you suggested. Could you please review them and let me know if any further improvements are needed?

**Comment by stevhliu** [comment]
Hey, there are still some unresolved comments!

**Comment by Tanuj-rai** [comment]
@stevhliu I've addressed all the comments. Please check to see if there's anything else that needs to be done.

File: docs/source/en/model_doc/granite.md [hunk 11]
```diff
@@ -9,62 +9,102 @@ Unless required by applicable law or agreed to in writing, software distributed
...
-In this paper, we study the correlation between optimal learning rate, batch size, and number of training tokens for the recently proposed WSD scheduler. After thousands of small experiments, we found a power-law relationship between variables and demonstrated its transferability across model sizes. Based on the observation, we propose a new learning rate scheduler, Power scheduler, that is agnostic about the number of training tokens and batch size. The experiment shows that combining the Power scheduler with Maximum Update Parameterization (\mup) can consistently achieve impressive performance with one set of hyperparameters regardless of the number of training tokens, batch size, model size, and even model architecture. Our 3B dense and MoE models trained with the Power scheduler achieve comparable performance as state-of-the-art small language models.
-We [open source](https://huggingface.co/collections/ibm/power-lm-66be64ae647ddf11b9808000) these pretrained models.*
+import torch
+from transformers import pipeline
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:44]: This is a duplicate

```suggestion
```

File: docs/source/en/model_doc/granite.md [hunk 12]
```diff
@@ -9,62 +9,102 @@ Unless required by applicable law or agreed to in writing, software distributed
...
+    device_map="auto",
+    attn_implementation="sdpa"
+)
+
+inputs = tokenizer("Explain quantum computing in simple terms", return_tensors="pt").to("cuda")
+outputs = model.generate(**inputs, max_length=50, cache_implementation="static")
+print(tokenizer.decode(outputs[0], skip_special_tokens=True))
+```
+</hfoption>
+<hfoption id="transformers-cli">
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:75]: The transformers cli recently got reworked!

```suggestion
<hfoption id="transformers CLI">
```

File: docs/source/en/model_doc/granite.md [hunk 13]
```diff
@@ -9,62 +9,102 @@ Unless required by applicable law or agreed to in writing, software distributed
...
+
+```python
+echo -e "Explain quantum computing simply." | transformers-cli run --task text-generation --model ibm-granite/granite-3.3-8b-instruct --device 0
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:78]:
```suggestion
echo -e "Explain quantum computing simply." | transformers run --task text-generation --model ibm-granite/granite-3.3-8b-instruct --device 0
```

File: docs/source/en/model_doc/granite.md [hunk 14]
```diff
@@ -9,62 +9,102 @@ Unless required by applicable law or agreed to in writing, software distributed
...
+    attn_implementation="sdpa",
+    quantization_config=quantization_config,
+)
+
+input_ids = tokenizer("Explain artificial intelligence to a 10 year old", return_tensors="pt").to("cuda")
+outputs = model.generate(**inputs, max_length=50, cache_implementation="static")
+print(tokenizer.decode(outputs[0], skip_special_tokens=True))
+```
 
+## Notes
```
Comment (stevhliu) [inline docs/source/en/model_doc/granite.md:106]:
```suggestion
```

File: docs/source/en/model_doc/granite.md (hunk 9 above)
Comment (stevhliu) [reply docs/source/en/model_doc/granite.md:88]: Still need to `import torch` and use the larger 8b checkpoint instead of the 2b one.

File: docs/source/en/model_doc/granite.md (hunk 4 above)
Comment (stevhliu) [reply docs/source/en/model_doc/granite.md:34]: Not resolved

**Review by stevhliu — COMMENTED** [review]
Thanks, still a few unresolved comments and need to update the `transformers cli` section

**Comment by Tanuj-rai** [comment]
Thank you @stevhliu, for reviewing and suggesting changes. I have committed them. Could you please take a look at it?

**Review by stevhliu — APPROVED** [review]
Thanks! Feel free to mark as ready and we can merge :)

**Comment by stevhliu** [comment]
Can you push an empty commit to trigger the `run_tests` CI test?

**Comment by Tanuj-rai** [comment]
Thank you so much @stevhliu for merging it.
